LoopKitUI/Views/FractionalQuantityPicker.swift

```diff
@@ -151,86 +154,21 @@ public struct FractionalQuantityPicker: View {
-        let valueTooLargeToSelect: Double = 1
-
-        switch whole {
-        case lowerAbsoluteBound.whole:
-            let smallestSelectableValue = fractionalValuesByWhole[whole]!.first!
-            let fractionalLowerRecommendedBound = lowerAbsoluteBound.whole == lowerRecommendedBound.whole
-                ? max(lowerRecommendedBound.fraction, smallestSelectableValue)
-                : valueTooLargeToSelect
-            let fractionalUpperRecommendedBound = lowerAbsoluteBound.whole == upperRecommendedBound.whole
-                ? upperRecommendedBound.fraction
-                : valueTooLargeToSelect
-            return Guardrail(
-                absoluteBounds: smallestSelectableValue...valueTooLargeToSelect,
-                recommendedBounds: fractionalLowerRecommendedBound...fractionalUpperRecommendedBound,
-                unit: unit
-            )
-        case ..<lowerRecommendedBound.whole:
-            return Guardrail(
-                absoluteBounds: valueTooSmallToSelect...valueTooLargeToSelect,
-                recommendedBounds: valueTooLargeToSelect...valueTooLargeToSelect,
-                unit: unit
-            )
-        case lowerRecommendedBound.whole:
-            let fractionalLowerRecommendedBound = lowerRecommendedBound.fraction + pow(10.0, -Double(Self.maximumSupportedPrecision))
-            return Guardrail(
-                absoluteBounds: valueTooSmallToSelect...valueTooLargeToSelect,
-                recommendedBounds: fractionalLowerRecommendedBound...valueTooLargeToSelect,
-                unit: unit
-            )
-        case ..<upperRecommendedBound.whole:
-            return nil
-        case upperRecommendedBound.whole:
-            return Guardrail(
-                absoluteBounds: valueTooSmallToSelect...valueTooLargeToSelect,
-                recommendedBounds: valueTooSmallToSelect...upperRecommendedBound.fraction,
-                unit: unit
-            )
-        case ..<upperAbsoluteBound.whole:
-            return Guardrail(
-                absoluteBounds: valueTooSmallToSelect...valueTooLargeToSelect,
-                recommendedBounds: valueTooSmallToSelect...valueTooSmallToSelect,
-                unit: unit
-            )
-        case upperAbsoluteBound.whole:
-            return Guardrail(
-                absoluteBounds: valueTooSmallToSelect...fractionalValuesByWhole[whole]!.last!,
-                recommendedBounds: valueTooSmallToSelect...valueTooSmallToSelect,
-                unit: unit
-            )
-        default:
-            assertionFailure("unreachable")
-            return nil
-        }
+    private func colorForFraction(_ fraction: Double) -> Color {
+        assert(fraction.fraction == fraction)
+
+        let valueIfFractionSelected = whole + fraction
+        let quantityIfFractionSelected = HKQuantity(unit: unit, doubleValue: valueIfFractionSelected)
+        return guardrail.color(for: quantityIfFractionSelected)
```

Comment:
:) so far this was much easier to follow.

Reply:
Awesome, my impression was the same. Wish this approach had occurred to me sooner!